src/Autofac/Core/IComponentRegistration.cs

```diff
@@ -76,60 +78,26 @@ public interface IComponentRegistration : IDisposable
+        /// Gets the resolve pipeline for the component.
         /// </summary>
-        event EventHandler<PreparingEventArgs> Preparing;
+        IResolvePipeline ResolvePipeline { get; }
 
         /// <summary>
-        /// Called by the container when an instance is required.
-        /// </summary>
-        /// <param name="context">The context in which the instance will be activated.</param>
-        /// <param name="service">The service being resolved.</param>
-        /// <param name="parameters">Parameters for activation. These may be modified by the event handler.</param>
-        [SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "1#", Justification = "The method may change the backing store of the parameter collection.")]
-        [SuppressMessage("Microsoft.Design", "CA1030:UseEventsWhereAppropriate", Justification = "This is the method that would raise the event.")]
-        void RaisePreparing(IComponentContext context, Service service, ref IEnumerable<Parameter> parameters);
-
-        /// <summary>
-        /// Fired when a new instance is being activated. The instance can be
-        /// wrapped or switched at this time by setting the Instance property in
-        /// the provided event arguments.
-        /// </summary>
-        event EventHandler<ActivatingEventArgs<object>> Activating;
-
-        /// <summary>
-        /// Called by the container once an instance has been constructed.
+        /// Gets a value indicating whether the registration is a 1:1 adapter on top
+        /// of another component (e.g., Meta, Func, or Owned).
         /// </summary>
-        /// <param name="context">The context in which the instance was activated.</param>
-        /// <param name="parameters">The parameters supplied to the activator.</param>
-        /// <param name="service">The service being resolved.</param>
-        /// <param name="instance">The instance.</param>
-        [SuppressMessage("Microsoft.Design", "CA1007:UseGenericsWhereAppropriate")]
-        [SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "2#", Justification = "The method may change the object as part of activation.")]
-        [SuppressMessage("Microsoft.Design", "CA1030:UseEventsWhereAppropriate", Justification = "This is the method that would raise the event.")]
-        void RaiseActivating(IComponentContext context, IEnumerable<Parameter> parameters, Service service, ref object instance);
+        bool IsAdapterForIndividualComponent { get; }
 
         /// <summary>
-        /// Fired when the activation process for a new instance is complete.
+        /// Provides an event that will be invoked just before a pipeline is built, and can be used to add additional middleware
+        /// at that point.
         /// </summary>
-        event EventHandler<ActivatedEventArgs<object>> Activated;
+        public event EventHandler<IResolvePipelineBuilder> PipelineBuilding;
```

Comment:
Do you mean an event that happens after all the default middleware has been added, but just prior to the actual pipeline build (i.e. the pipeline is still modifiable), or literally after the pipeline has been built into an `IResolvePipeline`?

For the former, there may be value to such an event, but if people add additional middleware after the defaults have been applied, the chance of messing up the overall behaviour increases significantly.

As for the latter, the only thing a built pipeline contains is an entrypoint method, and can only be used if you have a created `ResolveRequestContext`, which comes from the `ResolveOperation`. So I feel like that has a limited value.

Reply:
What I was thinking about was the symmetry of the thing - largely there are "before something happens" and "after something happens" events, so I wasn't sure if there was a need for an "after the pipeline is built" event. However, it does sound like it'd be pretty limited in value, possibly other than tracing or something I can't imagine what you'd do there.

I'm OK leaving it. I think we could add it later if someone requests it, and since it's highly unlikely anyone has implemented their own IComponentRegistration it'd just be a 0.1 version.